static/app/components/events/eventTags/eventTagsTree.tsx

```diff
@@ -0,0 +1,202 @@
+import styled from '@emotion/styled';
+
+import EventTagsContent from 'sentry/components/events/eventTags/eventTagContent';
+import {space} from 'sentry/styles/space';
+import type {EventTag} from 'sentry/types';
+import {generateQueryWithTag} from 'sentry/utils';
+import {useLocation} from 'sentry/utils/useLocation';
+import useOrganization from 'sentry/utils/useOrganization';
+
+const MAX_TREE_DEPTH = 4;
+interface TagTree {
+  [key: string]: TagTreeContent;
+}
+interface TagTreeContent {
+  subtree: TagTree;
+  value: string;
+  // These will be omitted on pseudo tags (see addToTagTree)
+  meta?: Record<any, any>;
+  originalTag?: EventTag;
+}
+
+function addToTagTree(
+  tree: TagTree,
+  tag: EventTag,
+  meta: Record<any, any>,
+  originalTag: EventTag
+): TagTree {
+  const branchMatches = tag.key.match(/\./g) ?? [];
+
+  const hasInvalidBranchCount =
+    branchMatches.length <= 0 || branchMatches.length > MAX_TREE_DEPTH;
+  const hasInvalidBranchSequence = (tag.key.match(/\.{2,}/g) ?? []).length > 0;
+
+  // Ignore tags with 0, or >4 branches, as well as sequential dots (e.g. 'some..tag')
+  if (hasInvalidBranchCount || hasInvalidBranchSequence) {
+    tree[tag.key] = {value: tag.value, subtree: {}, meta, originalTag};
+    return tree;
+  }
+  // E.g. 'device.model.version'
+  const splitIndex = tag.key.indexOf('.'); // 6
+  const trunk = tag.key.slice(0, splitIndex); // 'device'
+  const branch = tag.key.slice(splitIndex + 1); // 'model.version'
+
+  if (tree[trunk] === undefined) {
+    // We need to define the default as an empty space for alignment
+    tree[trunk] = {value: ' ', subtree: {}};
```

Comment:
This, in combination with `whitespace: pre` allow for the alignment of child tags when parents have no associated value. 

With this:

| parent | ' ' | 
| --- | --- |
| child | 'child-value-1' |

Without this:

| parent | 'child-value-1' | 
| --- | --- |
| child | |

Reply:
The problem with this is that we end up rendering text that can be highlighted that is just ' '. If anyone has any ideas on how to force the `<TagTreeValues/>` component to render even when it has no content, that would probably be better, that way everything continues to be aligned with its key.

Reply:
might be able to render an empty div or something instead and use grid to indent it?

Reply:
Was able to refactor this to use grid for indentation, also flattened the dom so its not as nested :+1: